File: packages/compose/src/main/kotlin/com/santi020k/lumen/SharedComponents.kt
package com.santi020k.lumen

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.ProgressBarRangeInfo
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.progressBarRangeInfo
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

enum class LumenCardVariant {
    Accent,
    Default,
    Destructive,
    Muted,
    Success,
    Warning
}

@Immutable
data class LumenCardPalette(
    val background: Color,
    val border: Color
)

fun lumenCardPalette(colors: LumenColorPalette, variant: LumenCardVariant): LumenCardPalette {
    if (variant == LumenCardVariant.Default) return LumenCardPalette(colors.surface, colors.line)

    if (variant == LumenCardVariant.Muted) return LumenCardPalette(colors.surfaceMuted, colors.line)

    val accent = when (variant) {
        LumenCardVariant.Accent -> colors.accent
        LumenCardVariant.Destructive -> colors.danger
        LumenCardVariant.Success -> colors.success
        LumenCardVariant.Warning -> colors.warning
        LumenCardVariant.Default, LumenCardVariant.Muted -> colors.line
    }

    return LumenCardPalette(
        background = accent.copy(alpha = 0.06f),
        border = accent.copy(alpha = 0.24f)
    )
}

@Composable
fun LumenCard(
    modifier: Modifier = Modifier,
    variant: LumenCardVariant = LumenCardVariant.Default,
    onClick: (() -> Unit)? = null,
    enabled: Boolean = true,
    content: @Composable () -> Unit
) {
    val colors = LocalLumenTheme.current.colors
    val palette = lumenCardPalette(colors, variant)
    val interactionModifier = if (onClick == null) {
        Modifier
    } else {
        Modifier.clickable(enabled = enabled, role = Role.Button, onClick = onClick)
    }

    Surface(
        modifier = modifier.then(interactionModifier),
        color = palette.background,
        contentColor = colors.ink,
        shape = RoundedCornerShape(LumenRadius.Lg),
        border = androidx.compose.foundation.BorderStroke(1.dp, palette.border)
    ) {
        Box(modifier = Modifier.padding(LumenSpacing.Xl)) {
            content()
        }
    }
}

enum class LumenAlertVariant {
    Default,
    Destructive,
    Success,
    Warning
}

private data class LumenAlertPalette(
    val background: Color,
    val border: Color,
    val foreground: Color
)

private fun alertPalette(
    colors: LumenColorPalette,
    variant: LumenAlertVariant
): LumenAlertPalette = when (variant) {
    LumenAlertVariant.Default -> LumenAlertPalette(colors.surface, colors.line, colors.ink)
    LumenAlertVariant.Destructive -> LumenAlertPalette(
        colors.danger.copy(alpha = 0.08f),
        colors.danger.copy(alpha = 0.36f),
        colors.danger
    )
    LumenAlertVariant.Success -> LumenAlertPalette(
        colors.success.copy(alpha = 0.08f),
        colors.success.copy(alpha = 0.36f),
        colors.success
    )
    LumenAlertVariant.Warning -> LumenAlertPalette(
        colors.warning.copy(alpha = 0.10f),
        colors.warning.copy(alpha = 0.42f),
        colors.warning
    )
}

@Composable
fun LumenAlert(
    modifier: Modifier = Modifier,
    variant: LumenAlertVariant = LumenAlertVariant.Default,
    content: @Composable () -> Unit
) {
    val palette = alertPalette(LocalLumenTheme.current.colors, variant)

    Surface(
        modifier = modifier,
        color = palette.background,
        contentColor = palette.foreground,
        shape = RoundedCornerShape(LumenRadius.Md),
        border = androidx.compose.foundation.BorderStroke(1.dp, palette.border)
    ) {
        Box(
            modifier = Modifier.padding(horizontal = LumenSpacing.Lg, vertical = LumenSpacing.Md)
        ) {
            content()
        }
    }
}

@Immutable
data class LumenProgressValue(
    val max: Float,
    val value: Float
) {
    val fraction: Float get() = value / max

    companion object {
        fun resolve(value: Float, max: Float): LumenProgressValue {
            val safeMax = if (max.isFinite() && max > 0f) max else 100f
            val finiteValue = if (value.isFinite()) value else 0f

            return LumenProgressValue(
                max = safeMax,
                value = finiteValue.coerceIn(0f, safeMax)
            )
        }
    }
}

@Composable
fun LumenProgress(
    value: Float,
    modifier: Modifier = Modifier,
    max: Float = 100f,
    label: String? = null
) {
    val colors = LocalLumenTheme.current.colors
    val progress = LumenProgressValue.resolve(value, max)
    val semanticsModifier = Modifier.semantics {
        progressBarRangeInfo = ProgressBarRangeInfo(progress.value, 0f..progress.max)
        if (label != null) contentDescription = label
    }

    Box(
        modifier = modifier
            .then(semanticsModifier)
            .fillMaxWidth()
            .height(LumenSpacing.Sm)
            .clip(RoundedCornerShape(LumenRadius.Full))
            .background(colors.surfaceStrong)
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(progress.fraction)
                .fillMaxHeight()
                .background(colors.brandSolid)
        )
    }
}

enum class LumenAvatarSize(val dimension: Dp) {
    Sm(32.dp),
    Md(40.dp),
    Lg(56.dp)
}

@Composable
fun LumenAvatar(
    modifier: Modifier = Modifier,
    painter: Painter? = null,
    fallback: String = "?",
    size: LumenAvatarSize = LumenAvatarSize.Md,
    label: String? = null
) {
    val colors = LocalLumenTheme.current.colors

    Box(
        modifier = modifier
            .size(size.dimension)
            .clip(CircleShape)
            .background(colors.surfaceMuted)
            .border(1.dp, colors.line, CircleShape)
            .clearAndSetSemantics {
                if (label != null) contentDescription = label
            },
        contentAlignment = Alignment.Center
    ) {
        if (painter == null) {
            Text(
                text = fallback,
                color = colors.ink,
                fontWeight = FontWeight.Bold
            )
        } else {
            Image(
                painter = painter,
                contentDescription = null,
                modifier = Modifier.fillMaxWidth().fillMaxHeight(),
                contentScale = ContentScale.Crop
            )
        }
    }
}

File: packages/swift/Sources/LumenUI/SharedComponents.swift
import SwiftUI

public enum LumenCardVariant: Sendable {
    case accent
    case `default`
    case destructive
    case muted
    case success
    case warning
}

public struct LumenCard<Content: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let action: (() -> Void)?
    private let content: Content
    private let padding: LumenSurfacePadding
    private let radius: LumenSurfaceRadius
    private let variant: LumenCardVariant

    public init(
        variant: LumenCardVariant = .default,
        padding: LumenSurfacePadding = .xl,
        radius: LumenSurfaceRadius = .lg,
        action: (() -> Void)? = nil,
        @ViewBuilder content: () -> Content
    ) {
        self.variant = variant
        self.padding = padding
        self.radius = radius
        self.action = action
        self.content = content()
    }

    public var body: some View {
        Group {
            if let action {
                Button(action: action) {
                    cardSurface
                }
                .buttonStyle(LumenCardButtonStyle())
            } else {
                cardSurface
            }
        }
    }

    private var cardSurface: some View {
        content
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(padding.value)
            .background(backgroundColor)
            .overlay {
                RoundedRectangle(cornerRadius: radius.value, style: .continuous)
                    .stroke(borderColor, lineWidth: 1)
            }
            .clipShape(RoundedRectangle(cornerRadius: radius.value, style: .continuous))
    }

    private var accentColor: Color? {
        switch variant {
        case .accent: theme.colors.accent
        case .default, .muted: nil
        case .destructive: theme.colors.danger
        case .success: theme.colors.success
        case .warning: theme.colors.warning
        }
    }

    private var backgroundColor: Color {
        switch variant {
        case .default: theme.colors.surface
        case .muted: theme.colors.surfaceMuted
        case .accent, .destructive, .success, .warning:
            accentColor?.opacity(0.06) ?? theme.colors.surface
        }
    }

    private var borderColor: Color {
        accentColor?.opacity(0.24) ?? theme.colors.line
    }
}

private struct LumenCardButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .opacity(configuration.isPressed ? 0.84 : 1)
            .scaleEffect(configuration.isPressed ? 0.995 : 1)
    }
}

public enum LumenAlertVariant: Sendable {
    case `default`
    case destructive
    case success
    case warning
}

public struct LumenAlert<Content: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let content: Content
    private let variant: LumenAlertVariant

    public init(
        variant: LumenAlertVariant = .default,
        @ViewBuilder content: () -> Content
    ) {
        self.variant = variant
        self.content = content()
    }

    public var body: some View {
        content
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, LumenSpacing.lg)
            .padding(.vertical, LumenSpacing.md)
            .foregroundStyle(foregroundColor)
            .background(backgroundColor)
            .overlay {
                RoundedRectangle(cornerRadius: LumenRadius.md, style: .continuous)
                    .stroke(borderColor, lineWidth: 1)
            }
            .clipShape(RoundedRectangle(cornerRadius: LumenRadius.md, style: .continuous))
    }

    private var accentColor: Color {
        switch variant {
        case .default: theme.colors.ink
        case .destructive: theme.colors.danger
        case .success: theme.colors.success
        case .warning: theme.colors.warning
        }
    }

    private var backgroundColor: Color {
        variant == .default ? theme.colors.surface : accentColor.opacity(0.08)
    }

    private var borderColor: Color {
        variant == .default ? theme.colors.line : accentColor.opacity(0.36)
    }

    private var foregroundColor: Color {
        variant == .default ? theme.colors.ink : accentColor
    }
}

public struct LumenProgressValue: Equatable, Sendable {
    public let max: Double
    public let value: Double

    public var fraction: Double { value / max }

    public static func resolve(value: Double, max: Double) -> LumenProgressValue {
        let safeMax = max.isFinite && max > 0 ? max : 100
        let finiteValue = value.isFinite ? value : 0
        let safeValue = min(safeMax, Swift.max(0, finiteValue))

        return LumenProgressValue(max: safeMax, value: safeValue)
    }
}

public struct LumenProgress: View {
    @Environment(\.lumenTheme) private var theme

    private let label: String?
    private let progress: LumenProgressValue

    public init(
        value: Double = 0,
        max: Double = 100,
        label: String? = nil
    ) {
        progress = LumenProgressValue.resolve(value: value, max: max)
        self.label = label
    }

    public var body: some View {
        GeometryReader { geometry in
            ZStack(alignment: .leading) {
                Capsule().fill(theme.colors.surfaceStrong)
                Capsule()
                    .fill(theme.colors.brandSolid)
                    .frame(width: geometry.size.width * progress.fraction)
            }
        }
        .frame(height: LumenSpacing.sm)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(label ?? "Progress"))
        .accessibilityValue(Text("\(Int(progress.fraction * 100)) percent"))
    }
}

public enum LumenAvatarSize: Sendable {
    case lg
    case md
    case sm

    var dimension: CGFloat {
        switch self {
        case .lg: 56
        case .md: 40
        case .sm: 32
        }
    }
}

public struct LumenAvatar: View {
    @Environment(\.lumenTheme) private var theme

    private let fallback: String
    private let image: Image?
    private let label: String?
    private let size: LumenAvatarSize

    public init(
        image: Image? = nil,
        fallback: String = "?",
        size: LumenAvatarSize = .md,
        label: String? = nil
    ) {
        self.image = image
        self.fallback = fallback
        self.size = size
        self.label = label
    }

    public var body: some View {
        ZStack {
            Circle().fill(theme.colors.surfaceMuted)

            if let image {
                image
                    .resizable()
                    .scaledToFill()
            } else {
                Text(fallback)
                    .font(fallbackFont)
                    .foregroundStyle(theme.colors.ink)
            }
        }
        .frame(width: size.dimension, height: size.dimension)
        .clipShape(Circle())
        .overlay {
            Circle().stroke(theme.colors.line, lineWidth: 1)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityHidden(label == nil)
        .accessibilityLabel(Text(label ?? ""))
    }

    private var fallbackFont: Font {
        size == .lg ? .headline.weight(.bold) : .callout.weight(.bold)
    }
}
